Model based testing

danbornside opened the pull request.

danbornside commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs:
```diff
@@ -637,6 +646,13 @@ data NewEpochState era = NewEpochState
   }
   deriving (Generic)
 
+nesEsLens :: Functor f
```
I think this is unused, and should be deleted

danbornside commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs:
```diff
@@ -482,6 +484,13 @@ type TransEpoch (c :: Type -> Constraint) era =
     c (Core.PParams era)
   )
 
+esLStateLens :: Functor f
```
I think this is unused and should be deleted.

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs:
```diff
@@ -40,13 +51,269 @@ import qualified Shelley.Spec.Ledger.STS.Bbody as STS
 -------------------------------------------------------------------------------}
 
+
+newtype ModelTxId = ModelTxId Integer
```
This is all in completely the wrong place... this is a testing concern

nc6 replied:
New testing stuff should live in the `cardano-ledger-test` package

danbornside replied:
These types are moved now.

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs:
```diff
@@ -40,13 +51,269 @@ import qualified Shelley.Spec.Ledger.STS.Bbody as STS
+data ModelBlock = ModelBlock
+  { _mbSlot :: !SlotNo
+  , _mbUtxo :: ![ModelTx]
+  }
```
Model looks reasonable, though

danbornside replied:
Can I resolve this?

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs:
```diff
@@ -40,13 +51,269 @@ import qualified Shelley.Spec.Ledger.STS.Bbody as STS
+  , ApplyBlockData era
+  )
+
+runApplyBlockData
```
I would like more documentation here!

danbornside replied:
i've removed this function.

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs:
```diff
@@ -40,13 +51,269 @@ import qualified Shelley.Spec.Ledger.STS.Bbody as STS
+    , STS (Core.EraRule "BBODY" era)
+    , Environment (Core.EraRule "TICK" era) ~ ()
+    , BaseM (Core.EraRule "TICK" era) ~ BaseM (Core.EraRule "BBODY" era)
+    ) => ApplyBlock' era where
```
I still don't understand the point of this, or why it's needed. We should be testing the existing API

danbornside replied:
I've combined the two classes and specialised these types to ShelleyBase, NewEraState, SlotNo, and Block; as appropriate.

nc6 commented on shelley/chain-and-ledger/shelley-spec-ledger-test/shelley-spec-ledger-test.cabal:
```diff
@@ -44,8 +44,10 @@ library
                       test
   exposed-modules:
     Test.Cardano.Crypto.VRF.Fake
+    Test.Shelley.Spec.Ledger.ApplyBlock
```
I'd like the new testing code to live in `cardano-ledger-test` - it should be general across eras, not stuck in the shelley era.

HectorAE replied:
~~Looks like that package is currently defined under the `byron` subdirectory, so we might want to move it elsewhere too, perhaps to the top level?~~

Realized this suggestion has already been carried out in the `master` branch.

nc6 commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/ApplyBlock.hs:
```diff
@@ -0,0 +1,195 @@
+
+type KeyPair' crypto = (KeyPair 'Payment crypto, KeyPair 'Staking crypto)
+
+data ShelleyEraState crypto = ShelleyEraState
```
I'm a bit concerned that this is tied to the Shelley era. Given this is basically model state, we shouldn't need it to be era specific.

danbornside replied:
This was not well named and has been renamed.

nc6 commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/ApplyBlock.hs:
```diff
@@ -0,0 +1,195 @@
+    ( PraosCrypto crypto
+    , KES.Signable (KES crypto) ~ SignableRepresentation
+    , DSIGN.Signable (DSIGN crypto) ~ SignableRepresentation
+    ) => TraceApplyBlock (ShelleyEra crypto) where
```
So is this the core interface that must be implemented per era?

danbornside replied:
That's correct.  

Can I resolve this?

nc6 commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Example.hs:
```diff
@@ -0,0 +1,252 @@
+
+data ModelChain a
+
+data ModelChainPredicateFailure a
```
I don't understand what this is and why it's living in `Example`

danbornside replied:
I never intended to commit this in the first place and have deleted it.

JaredCorduan commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxWitness.hs:
```diff
@@ -325,3 +332,32 @@ instance
   where
   prettyA = ppTxWitness
+
+instance Era era => Semigroup (Redeemers era) where
```
Unfortunately, this Semigroup instance is unlawful and perhaps a bit dangerous. The `Redeemers` type is one of the types that lazily stores its own serialization:
```
newtype Redeemers era = RedeemersConstr (MemoBytes (RedeemersRaw era))
```
So imagine this scenario:
A user out in userland submits a transaction with a redeemer `r`, and they way they choose to cbor encode the redeemer is _not_ what our code would do (but is still legal cbor). Since `r` stores the original bytes of the redeemer, this is what we will use when we need a hash of `r` in the ledger rules. This way the rules and the user can agree to the proper hash of `r`. However, the hash of `r <> mempty` will be different! Why? `r <> mempty` will create a new `Redeemer` using the `Redeemer` pattern, and so the lazy serialization field becomes a thunk. When we later want the hash of `r <> mempty`, we will then serialize the map ourselves, possibly making a different choice for serialization than the user.

JaredCorduan replied:
erm, I guess it's the Monoid instance that has the broken law, but it's `<>` that is causing the grief :)

danbornside replied:
hmm.  The only reason I wrote this instance is to have a "blank" redemer to add witness to.  I was a bit concerned when I wrote the instance.

danbornside replied:
I guess the only valid thing to do is use the pattern constructor at the site I create the redeemer in the elaborator.

JaredCorduan replied:
would a `Default` instance be convenient? We use those in several places, and that won't cause any problems.

danbornside committed:
implements model based testing framework.

- generalise the ApplyBlock class a bit to take ApplySTSOpts for use by testing.
- add a class EraElaborator to apply era-agnostic blockchain model interactions
  to real ledger implementations
- instance of EraElaborator for Alonzo as Shelley eras.
- simple unit test which applies simple transactions using EraElaborator to any
  era.
- add some delegation related features to the model
- unhide mempool api.

JaredCorduan commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Translation.hs:
```diff
@@ -205,7 +206,7 @@ translateTxOut (Shelley.TxOutCompact addr value) =
 
 translatePParams ::
-  AlonzoGenesis -> Shelley.PParams (MaryEra c) -> PParams (AlonzoEra c)
+  AlonzoGenesis -> Shelley.PParams era -> PParams (AlonzoEra c)
```
was this needed?

danbornside replied:
I'd written an instance of "CanStartFromGenesis" that needed it, but should be out from after the rebase, I'll see if this can be backed out'  on the other hand, it's not at all clear to me why it needs to specify `MaryEra` and not be fully general.  PParams is not a type family and has `era` as a phantom parameter.

JaredCorduan replied:
thanks! it's more that I'm not ever expecting to need to translate from any era besides Mary, and I was curious why

JaredCorduan commented on cardano-ledger-test/src/Test/Cardano/Ledger/Elaborators/Alonzo.hs:
```diff
@@ -0,0 +1,87 @@
+    , maxTxExUnits = ExUnits 100 100
+    , maxBlockExUnits = ExUnits 100 100
+    , maxValSize = 100000
+    , collateralPercentage = 1 -- TODO: what is a normal value for this?
```
100 or 150 would both be pretty normal.

JaredCorduan commented on cardano-ledger-test/src/Test/Cardano/Ledger/Elaborators/Alonzo.hs:
```diff
@@ -0,0 +1,87 @@
+    , maxBlockExUnits = ExUnits 100 100
+    , maxValSize = 100000
+    , collateralPercentage = 1 -- TODO: what is a normal value for this?
+    , maxCollateralInputs = 100 -- TODO: what is a normal value for this?
```
100 is fine. I think folks mentioned 10 as a possibility for mainnet.

JaredCorduan commented on cardano-ledger-test/src/Test/Cardano/Ledger/Elaborators/Alonzo.hs:
```diff
@@ -0,0 +1,87 @@
+
+instance Default AlonzoGenesis where
+  def = AlonzoGenesis
+    { adaPerUTxOWord = Coin 10000  -- error "Default AlonzoGenesis :: adaPerUTxOWord"
```
I don't understand the comment, sorry

JaredCorduan commented on cardano-ledger-test/src/Test/Cardano/Ledger/ModelChain.hs:
```diff
@@ -0,0 +1,698 @@
+        k = TestKeyPair
+          { _tkpKeyPair = keyPair
+          , _tkpVRF = vrf
+          , _tkpAddr = toAddr Testnet keyPair -- TODO: network can be read from Globals
```
it's fine by me to leave this hard coded as `Testnet`.

danbornside replied:
My thinking was that the network is in Globals; and getting it from context is a way to avoid some latent bugs.

JaredCorduan replied:
yea, that's reasonable

JaredCorduan commented on cardano-ledger-test/src/Test/Cardano/Ledger/ModelChain/Properties.hs:
```diff
@@ -0,0 +1,158 @@
+              , ModelTxOut "bob" 8999820000000
+              ]
+            , _mtxFee = 100_000_000_000
+            , _mtxWdrl = Map.singleton "alice" (Coin 8999820000000)
```
I'm confused what's going on here. I see Bob getting `8999820000000` in an output. And that same number used for what look like a reward withdrawal for Alice. Are these related, and how do you know how much Alice can withdraw?

danbornside replied:
i guess the way to interpret this is that "alice" is gifting "bob" with the rewards, the addresses used aren't meaningfully distinct entities, only distinct keys.  There's no significance to the addresses being different.

The value is what happens to be rewarded in this exact sequence of inputs.  The hardcoded value goes away in another branch ( https://github.com/obsidiansystems/cardano-ledger-specs/compare/model-based-testing...obsidiansystems:model-based-testing-ma?expand=1 )

JaredCorduan replied:
ah!

JaredCorduan commented on cardano-ledger-test/test/Tests.hs:
```diff
@@ -30,7 +31,8 @@ mainTests =
       alonzoBBODYexamples,
-      collectOrderingAlonzo
+      collectOrderingAlonzo,
+      newUnitTests
```
model unit tests?

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs:
```diff
@@ -67,6 +72,44 @@ class
+  -- | Apply the header level ledger transition.
+  --
+  -- This handles checks and updates that happen on a slot tick, as well as a
+  -- few header level checks, such as size constraints.
```
so this is just `TICK`, but where you can supply your own options, right? Maybe a comment to that effect? (and similarly for `applyBlockOpts` below)

JaredCorduan reviewed (approved):
This looks great to me! Thanks @danbornside !

JaredCorduan commented on cardano-ledger-test/src/Test/Cardano/Ledger/ModelChain.hs:
```diff
@@ -0,0 +1,698 @@
+import qualified Shelley.Spec.Ledger.UTxO as UTxO
+import qualified Shelley.Spec.Ledger.TxBody as Shelley
+import Cardano.Ledger.SafeHash (SafeHash)
+
```
There are a lot of open imports here, I always prefer to use either explicit imports or qualified ones. That may just be me, though. You don't necessarily have to do this, I can maybe do it later.

nc6 commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Presets.hs:
```diff
@@ -18,6 +18,7 @@ module Test.Shelley.Spec.Ledger.Generator.Presets
     genesisDelegs0,
     someKeyPairs,
     keyPairs,
+    issuerKeys,
```
This change needed?

nc6 reviewed (approved):
Looks generally good. I'd appreciate some more documentation in various parts!

danbornside committed:
Fixup re PR comments

danbornside committed:
add some documentation, remove an unused abstraction.

JaredCorduan commented:
replaced by #2338
